File: packages/client/src/query.ts
import {
  type Condition,
  type DurationParts,
  type RelativeNow,
  normalizeCondition,
} from "./condition";
import type { ListQuery, ListResponse } from "./types";

/**
 * Type-safe fluent query builder — a Drizzle/Supabase-style ergonomics layer
 * that COMPILES to the canonical JSON `Condition` / `ListQuery` the REST API
 * already speaks. It is NOT a new wire format: `.toQuery()` returns a plain
 * `ListQuery`, so everything (permissions, AI plans, serialization) stays on
 * the one JSON grammar.
 *
 *   const { data } = await client.from<Order>("orders").query()
 *     .where(f => f.and(
 *       f.eq("status", "active"),
 *       f.gte("total", 100),
 *       f.rel("customer", c => c.eq("tier", "gold")),   // → "customer.tier"
 *       f.gte("placed_at", f.now({ sub: { months: 1 } })),
 *     ))
 *     .select("id", "total", "customer.name")
 *     .orderBy("-placed_at", "id")
 *     .limit(50)
 *     .list();
 *
 * Field args are typed `keyof T | (string & {})` — autocomplete for known
 * columns, dotted relation paths still allowed, no codegen required.
 */

/** A known column of `T`, or any string (dotted relation paths, computed). */
export type FieldKey<T> = (keyof T & string) | (string & {});

/** `field` ascending, or `-field` descending. */
export type SortKey<T> = FieldKey<T> | `-${string}`;

/** Prefix every leaf field key of a condition with `head.` (relation hop). */
const prefixKeys = (cond: Condition, head: string): Condition => {
  const c = cond as Record<string, unknown>;
  if (Array.isArray(c.$and)) {
    return { $and: (c.$and as Condition[]).map((x) => prefixKeys(x, head)) };
  }
  if (Array.isArray(c.$or)) {
    return { $or: (c.$or as Condition[]).map((x) => prefixKeys(x, head)) };
  }
  if (c.$not !== undefined) {
    return { $not: prefixKeys(c.$not as Condition, head) };
  }
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(c)) out[`${head}.${k}`] = v;
  return out as Condition;
};

/** Fluent condition factory passed to `QueryBuilder.where(...)`. Each method
 *  returns a canonical {@link Condition}; multiple are combined with `and`/`or`. */
export interface FilterBuilder<T> {
  /** `field = value`. */
  eq(field: FieldKey<T>, value: unknown): Condition;
  /** `field != value`. */
  neq(field: FieldKey<T>, value: unknown): Condition;
  /** `field > value`. */
  gt(field: FieldKey<T>, value: unknown): Condition;
  /** `field >= value`. */
  gte(field: FieldKey<T>, value: unknown): Condition;
  /** `field < value`. */
  lt(field: FieldKey<T>, value: unknown): Condition;
  /** `field <= value`. */
  lte(field: FieldKey<T>, value: unknown): Condition;
  /** `field IN (values)`. */
  in(field: FieldKey<T>, values: unknown[]): Condition;
  /** `field NOT IN (values)`. */
  nin(field: FieldKey<T>, values: unknown[]): Condition;
  /** `field BETWEEN lo AND hi` (inclusive). */
  between(field: FieldKey<T>, lo: unknown, hi: unknown): Condition;
  /** `field IS NULL` (or `IS NOT NULL` when `isNull` is false). */
  isNull(field: FieldKey<T>, isNull?: boolean): Condition;
  /** `field` is NULL or empty string. */
  empty(field: FieldKey<T>): Condition;
  /** `field` is neither NULL nor empty string. */
  nempty(field: FieldKey<T>): Condition;
  /** `field` contains the substring `value`. */
  contains(field: FieldKey<T>, value: string): Condition;
  /** Case-insensitive {@link contains}. */
  icontains(field: FieldKey<T>, value: string): Condition;
  /** `field` starts with `value`. */
  startsWith(field: FieldKey<T>, value: string): Condition;
  /** `field` ends with `value`. */
  endsWith(field: FieldKey<T>, value: string): Condition;
  /** Logical AND of the given conditions. */
  and(...conds: Condition[]): Condition;
  /** Logical OR of the given conditions. */
  or(...conds: Condition[]): Condition;
  /** Logical NOT of a condition. */
  not(cond: Condition): Condition;
  /** Traverse a relation: keys produced by `build` are prefixed with `head.`. */
  rel<R = Record<string, unknown>>(
    head: FieldKey<T>,
    build: (f: FilterBuilder<R>) => Condition,
  ): Condition;
  /** Relative-date value, e.g. `f.now({ sub: { months: 1 } })`. */
  now(opts?: { add?: DurationParts; sub?: DurationParts }): RelativeNow;
}

const makeFilterBuilder = <T>(): FilterBuilder<T> => {
  const leaf = (field: string, op: string, value: unknown): Condition =>
    ({ [field]: { [op]: value } }) as Condition;
  return {
    eq: (f, v) => leaf(f, "_eq", v),
    neq: (f, v) => leaf(f, "_neq", v),
    gt: (f, v) => leaf(f, "_gt", v),
    gte: (f, v) => leaf(f, "_gte", v),
    lt: (f, v) => leaf(f, "_lt", v),
    lte: (f, v) => leaf(f, "_lte", v),
    in: (f, v) => leaf(f, "_in", v),
    nin: (f, v) => leaf(f, "_nin", v),
    between: (f, lo, hi) => leaf(f, "_between", [lo, hi]),
    isNull: (f, isNull = true) => leaf(f, "_null", isNull),
    empty: (f) => leaf(f, "_empty", true),
    nempty: (f) => leaf(f, "_nempty", true),
    contains: (f, v) => leaf(f, "_contains", v),
    icontains: (f, v) => leaf(f, "_icontains", v),
    startsWith: (f, v) => leaf(f, "_starts_with", v),
    endsWith: (f, v) => leaf(f, "_ends_with", v),
    and: (...conds) => ({ $and: conds }),
    or: (...conds) => ({ $or: conds }),
    not: (cond) => ({ $not: cond }),
    rel: (head, build) =>
      prefixKeys(build(makeFilterBuilder()), head as string),
    now: (opts = {}) => ({ $now: { ...opts } }),
  };
};

/** Fluent, type-safe query builder returned by `from(slug).query()`. Chain
 *  `.where`/`.select`/`.orderBy`/… and finish with `.list()` (or `.toQuery()`). */
export class QueryBuilder<T extends Record<string, unknown>> {
  private _filter?: Condition;
  private _sort: string[] = [];
  private _fields: string[] = [];
  private _expand: string[] = [];
  private _limit?: number;
  private _offset?: number;
  private _cursor?: string;
  private _meta?: "filter_count" | "total_count" | "*";
  private _locale?: string;
  private _q?: string;

  constructor(private readonly listFn: (q: ListQuery) => Promise<ListResponse<T>>) {}

  /** Build the filter fluently with a {@link FilterBuilder}. */
  where(build: (f: FilterBuilder<T>) => Condition): this {
    this._filter = normalizeCondition(build(makeFilterBuilder<T>()));
    return this;
  }
  /** Replace the filter with a raw canonical condition (escape hatch). */
  filter(cond: Condition): this {
    this._filter = normalizeCondition(cond);
    return this;
  }
  /** Project only these fields (column allow-list). */
  select(...fields: FieldKey<T>[]): this {
    this._fields.push(...(fields as string[]));
    return this;
  }
  /** Sort by one or more keys (`"-field"` for descending). */
  orderBy(...sorts: SortKey<T>[]): this {
    this._sort.push(...(sorts as string[]));
    return this;
  }
  /** Inline single-hop relations (replaces each FK with the related object). */
  expand(...rels: FieldKey<T>[]): this {
    this._expand.push(...(rels as string[]));
    return this;
  }
  /** Project `localized` fields to one locale, or `"*"` for the full map. */
  locale(loc: string): this {
    this._locale = loc;
    return this;
  }
  /** Free-text search across readable text fields. */
  search(text: string): this {
    this._q = text;
    return this;
  }
  /** Max rows to return. */
  limit(n: number): this {
    this._limit = n;
    return this;
  }
  /** Number of rows to skip (paging). */
  offset(n: number): this {
    this._offset = n;
    return this;
  }
  /** Keyset (seek) pagination — pass `""` for the first page, then echo each
   *  response's `next_cursor`. Flat latency at any depth; supersedes `offset`. */
  cursor(c: string): this {
    this._cursor = c;
    return this;
  }
  /** Ask the server for a count alongside the page (`filter_count` / `total_count`). */
  withMeta(m: "filter_count" | "total_count" | "*"): this {
    this._meta = m;
    return this;
  }
  /** Assemble the plain `ListQuery` — the canonical JSON the REST API takes. */
  toQuery(): ListQuery {
    const q: ListQuery = {};
    if (this._filter) q.filter = this._filter;
    if (this._sort.length) q.sort = this._sort;
    if (this._fields.length) q.fields = this._fields;
    if (this._expand.length) q.expand = this._expand;
    if (this._limit !== undefined) q.limit = this._limit;
    if (this._offset !== undefined) q.offset = this._offset;
    if (this._cursor !== undefined) q.cursor = this._cursor;
    if (this._meta) q.meta = this._meta;
    if (this._locale) q.locale = this._locale;
    if (this._q) q.q = this._q;
    return q;
  }
  /** Run the query and return the page of rows. */
  list(): Promise<ListResponse<T>> {
    return this.listFn(this.toQuery());
  }
}

File: sdks/python/src/backlex/query.py
"""Fluent query builder + filter normalization.

A Python port of ``packages/client/src/query.ts`` and the schema-blind half of
``packages/core/src/condition.ts``. It is **not** a new wire format: every
builder compiles to the same canonical JSON ``Condition`` / ``ListQuery`` the
REST API already speaks, so permissions, AI plans, and serialization all stay on
the one grammar.

    rows = (
        client.from_("orders").query()
        .where(lambda f: f.and_(
            f.eq("status", "active"),
            f.gte("total", 100),
            f.rel("customer", lambda c: c.eq("tier", "gold")),  # -> "customer.tier"
            f.gte("placed_at", f.now(sub={"months": 1})),
        ))
        .select("id", "total", "customer.name")
        .order_by("-placed_at", "id")
        .limit(50)
        .list()
    )["data"]
"""

from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional, cast

from .types import Condition, ListQuery, ListResponse

# ---------------------------------------------------------------------------
# normalize_condition — schema-blind subset (matches the SDK's usage, which
# never passes ``relationFields``: the builder already emits dotted keys).
# ---------------------------------------------------------------------------


def _is_plain_object(v: Any) -> bool:
    return isinstance(v, dict)


def _looks_like_comparison(o: Dict[str, Any]) -> bool:
    keys = list(o.keys())
    return len(keys) > 0 and all(k.startswith("_") for k in keys)


def normalize_condition(raw: Any) -> Condition:
    """Turn any accepted filter shape into the canonical ``Condition``.

    Handles ``$and`` / ``$or`` / ``$not`` (and their ``_`` aliases) and implicit
    equality (``{"status": "active"}`` -> ``{"status": {"_eq": "active"}}``).
    Idempotent. Non-dict input is returned unchanged.
    """
    if not _is_plain_object(raw):
        return cast(Condition, raw)

    and_ = raw.get("$and", raw.get("_and"))
    if isinstance(and_, list):
        return {"$and": [normalize_condition(c) for c in and_]}
    or_ = raw.get("$or", raw.get("_or"))
    if isinstance(or_, list):
        return {"$or": [normalize_condition(c) for c in or_]}
    not_ = raw.get("$not", raw.get("_not"))
    if not_ is not None:
        return {"$not": normalize_condition(not_)}

    out: Dict[str, Any] = {}
    for key, value in raw.items():
        if _is_plain_object(value) and _looks_like_comparison(value):
            out[key] = value
        elif _is_plain_object(value):
            # Unknown object shape (json literal, schema-blind nesting) — pass through.
            out[key] = value
        else:
            out[key] = {"_eq": value}
    return out


def _prefix_keys(cond: Condition, head: str) -> Condition:
    """Prefix every leaf field key of a condition with ``head.`` (relation hop)."""
    if isinstance(cond.get("$and"), list):
        return {"$and": [_prefix_keys(x, head) for x in cond["$and"]]}
    if isinstance(cond.get("$or"), list):
        return {"$or": [_prefix_keys(x, head) for x in cond["$or"]]}
    if cond.get("$not") is not None:
        return {"$not": _prefix_keys(cond["$not"], head)}
    return {f"{head}.{k}": v for k, v in cond.items()}


# ---------------------------------------------------------------------------
# FilterBuilder — the ``f`` passed to ``.where(lambda f: ...)``.
# Python keywords (and/or/not/in) get a trailing underscore.
# ---------------------------------------------------------------------------


class FilterBuilder:
    """Leaf + logical condition constructors. Each method returns a ``Condition``."""

    @staticmethod
    def _leaf(field: str, op: str, value: Any) -> Condition:
        return {field: {op: value}}

    def eq(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_eq", value)

    def neq(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_neq", value)

    def gt(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_gt", value)

    def gte(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_gte", value)

    def lt(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_lt", value)

    def lte(self, field: str, value: Any) -> Condition:
        return self._leaf(field, "_lte", value)

    def in_(self, field: str, values: List[Any]) -> Condition:
        return self._leaf(field, "_in", values)

    def nin(self, field: str, values: List[Any]) -> Condition:
        return self._leaf(field, "_nin", values)

    def between(self, field: str, lo: Any, hi: Any) -> Condition:
        return self._leaf(field, "_between", [lo, hi])

    def is_null(self, field: str, is_null: bool = True) -> Condition:
        return self._leaf(field, "_null", is_null)

    def empty(self, field: str) -> Condition:
        return self._leaf(field, "_empty", True)

    def nempty(self, field: str) -> Condition:
        return self._leaf(field, "_nempty", True)

    def contains(self, field: str, value: str) -> Condition:
        return self._leaf(field, "_contains", value)

    def icontains(self, field: str, value: str) -> Condition:
        return self._leaf(field, "_icontains", value)

    def starts_with(self, field: str, value: str) -> Condition:
        return self._leaf(field, "_starts_with", value)

    def ends_with(self, field: str, value: str) -> Condition:
        return self._leaf(field, "_ends_with", value)

    def and_(self, *conds: Condition) -> Condition:
        return {"$and": list(conds)}

    def or_(self, *conds: Condition) -> Condition:
        return {"$or": list(conds)}

    def not_(self, cond: Condition) -> Condition:
        return {"$not": cond}

    def rel(self, head: str, build: "Callable[[FilterBuilder], Condition]") -> Condition:
        """Traverse a relation: keys produced by ``build`` are prefixed ``head.``."""
        return _prefix_keys(build(FilterBuilder()), head)

    def now(
        self,
        add: Optional[Dict[str, int]] = None,
        sub: Optional[Dict[str, int]] = None,
    ) -> Dict[str, Any]:
        """Relative-date value, e.g. ``f.now(sub={"months": 1})``."""
        opts: Dict[str, Any] = {}
        if add is not None:
            opts["add"] = add
        if sub is not None:
            opts["sub"] = sub
        return {"$now": opts}


ListFn = Callable[[ListQuery], ListResponse]


class QueryBuilder:
    """Chainable assembler that compiles to a plain ``ListQuery``."""

    def __init__(self, list_fn: ListFn) -> None:
        self._list_fn = list_fn
        self._filter: Optional[Condition] = None
        self._sort: List[str] = []
        self._fields: List[str] = []
        self._limit: Optional[int] = None
        self._offset: Optional[int] = None
        self._meta: Optional[str] = None

    def where(self, build: Callable[[FilterBuilder], Condition]) -> "QueryBuilder":
        self._filter = normalize_condition(build(FilterBuilder()))
        return self

    def filter(self, cond: Condition) -> "QueryBuilder":
        """Replace the filter with a raw canonical condition (escape hatch)."""
        self._filter = normalize_condition(cond)
        return self

    def select(self, *fields: str) -> "QueryBuilder":
        self._fields.extend(fields)
        return self

    def order_by(self, *sorts: str) -> "QueryBuilder":
        self._sort.extend(sorts)
        return self

    def limit(self, n: int) -> "QueryBuilder":
        self._limit = n
        return self

    def offset(self, n: int) -> "QueryBuilder":
        self._offset = n
        return self

    def with_meta(self, m: str) -> "QueryBuilder":
        """Request an extra COUNT: ``"filter_count"``, ``"total_count"``, or ``"*"``."""
        self._meta = m
        return self

    def to_query(self) -> ListQuery:
        """Assemble the plain ``ListQuery`` — the canonical JSON the REST API takes."""
        q: ListQuery = {}
        if self._filter:
            q["filter"] = self._filter
        if self._sort:
            q["sort"] = self._sort
        if self._fields:
            q["fields"] = self._fields
        if self._limit is not None:
            q["limit"] = self._limit
        if self._offset is not None:
            q["offset"] = self._offset
        if self._meta:
            q["meta"] = self._meta
        return q

    def list(self) -> ListResponse:
        return self._list_fn(self.to_query())
